ddl/worker.go

```diff
@@ -0,0 +1,278 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/model"
+)
+
+func (d *ddl) startJob(ctx context.Context, job *model.Job) error {
+	// for every DDL, we must commit current transaction.
+	if err := ctx.FinishTxn(false); err != nil {
+		return errors.Trace(err)
+	}
+
+	// Create a new job and queue it.
+	err := d.meta.RunInNewTxn(false, func(m *meta.TMeta) error {
+		var err error
+		job.ID, err = m.GenGlobalID()
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		err = m.EnQueueDDLJob(job)
+		return errors.Trace(err)
+	})
+
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// notice worker that we push a new job and wait the job done.
+	asyncNotify(d.jobCh)
+
+	jobID := job.ID
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-d.jobDoneCh:
+		case <-ticker.C:
+		}
+
+		job, err = d.getHistoryJob(jobID)
+		if err != nil {
+			log.Errorf("get history job err %v, check again", err)
+			continue
+		} else if job == nil {
+			log.Warnf("job %d is not in history, maybe not run", jobID)
+			continue
+		}
+
+		// if a job is a history table, the state must be JobDone or JobCancel.
+		if job.State == model.JobDone {
+			return nil
+		}
+
+		return errors.Errorf("job is %s, err :%v", job.State, job.Error)
+	}
+}
+
+func (d *ddl) getHistoryJob(id int64) (*model.Job, error) {
+	var job *model.Job
+
+	err := d.meta.RunInNewTxn(false, func(t *meta.TMeta) error {
+		var err1 error
+		job, err1 = t.GetHistoryDDLJob(id)
+		return errors.Trace(err1)
+	})
+
+	return job, errors.Trace(err)
+}
+
+func asyncNotify(ch chan struct{}) {
+	select {
+	case ch <- struct{}{}:
+	default:
+	}
+}
+
+func (d *ddl) verifyOwner(t *meta.TMeta) error {
+	owner, err := t.GetDDLOwner()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if owner == nil {
+		owner = &model.Owner{}
+		// try to set onwer
+		owner.OwnerID = d.uuid
+	}
+
+	now := time.Now().Unix()
+	maxTimeout := int64(4 * d.lease)
+	if owner.OwnerID == d.uuid || now-owner.LastUpdateTS > maxTimeout {
+		owner.OwnerID = d.uuid
+		owner.LastUpdateTS = now
+
+		// update or try to set itself as owner.
+		err = t.SetDDLOwner(owner)
+	}
+
+	if err != nil {
+		return errors.Trace(err)
+	} else if owner.OwnerID != d.uuid {
+		return errors.Trace(ErrNotOwner)
+	}
+
+	return nil
+}
+
+// every time we enter another state, we must call this function.
+func (d *ddl) updateJob(t *meta.TMeta, j *model.Job) error {
```

Comment:
I prefer using job instead of j as variable which is the same as startJob function.